src/Runtime/Distributions/Automata/Automaton.Determinization.cs

```diff
@@ -124,110 +250,121 @@ public bool TryDeterminize()
-                    this.stateIdToWeight.TryGetValue(stateId, out weight);
+                public int Count =>
+                    this.weightedStates.IsNull
+                        ? 1
+                        : this.weightedStates.Count;
+
+                public WeightedState this[int index] =>
+                    this.weightedStates.IsNull
+                        ? new WeightedState(this.singleStateIndex, Weight.One)
+                        : this.weightedStates[index];
 
                 /// <summary>
-                /// Checks whether the state with a given id is present in the set.
+                /// Checks whether this object is equal to a given one.
                 /// </summary>
-                /// <param name="stateId">The state id,</param>
+                /// <param name="obj">The object to compare this object with.</param>
                 /// <returns>
-                /// <see langword="true"/> if the given state id was present in the set,
+                /// <see langword="true"/> if the objects are equal,
                 /// <see langword="false"/> otherwise.
                 /// </returns>
-                public bool ContainsState(int stateId) =>
-                    this.stateIdToWeight.ContainsKey(stateId);
+                public override bool Equals(object obj) => obj is WeightedStateSet that && this.Equals(that);
 
                 /// <summary>
                 /// Checks whether this object is equal to a given one.
                 /// </summary>
-                /// <param name="obj">The object to compare this object with.</param>
+                /// <param name="that">The object to compare this object with.</param>
                 /// <returns>
                 /// <see langword="true"/> if the objects are equal,
                 /// <see langword="false"/> otherwise.
                 /// </returns>
-                public override bool Equals(object obj)
+                public bool Equals(WeightedStateSet that)
                 {
-                    if (obj == null || obj.GetType() != typeof(WeightedStateSet))
+                    if (this.Count != that.Count)
                     {
                         return false;
                     }
 
-                    var other = (WeightedStateSet)obj;
-
-                    if (this.stateIdToWeight.Count != other.stateIdToWeight.Count)
+                    for (var i = 0; i < this.Count; ++i)
                     {
-                        return false;
-                    }
-
-                    foreach (KeyValuePair<int, Weight> pair in this.stateIdToWeight)
-                    {
-                        // TODO: Should we allow for some tolerance? But what about hashing then?
-                        Weight otherWeight;
-                        if (!other.stateIdToWeight.TryGetValue(pair.Key, out otherWeight) || otherWeight != pair.Value)
+                        var state1 = this[i];
```

Comment:
If you had the same set but in a different order, this would fail. Is that desired behaviour?

Reply:
I see later that your WeightedStateSetBuilder Get method orders the weighted set.  Is there an assumption that this is always the mechanism by which weight sets are built?

Reply:
Yes, this is an assumption. I will asserts in debug build that this is the case